### PR TITLE
NAS-109925 / 21.06 / Allow specifying extra arguments with id keyword in rest api

### PR DESCRIPTION
**Pull request by sonicaj**
This PR adds following changes:

1) Updates documentation to not specify filterable arguments like `sort` when using `id` keyword in rest api endpoint.
2) Allows specifying `query-options.extra` arguments when using `id` keyword.
3) Fixes an issue which correctly outlines that `query-options.extra` is supported in request body with `id` keyword in rest api endpoint and not the complete `filterable` request body.

### CODE REVIEW

File: src/middlewared/middlewared/restful.py [hunk 1]
```diff
@@ -497,9 +511,31 @@ def get_post_args(self, req, resp, **kwargs):
                     filterid = kwargs['id']
                     if filterid.isdigit():
                         filterid = int(filterid)
+                    extra = {}
+                    text = await req.text()
```
Comment (themylogin) [inline src/middlewared/middlewared/restful.py:515]: Are we reading request body of a GET method?

**Review by themylogin — CHANGES_REQUESTED** [review]

File: src/middlewared/middlewared/restful.py (hunk 1 above)
Comment (sonicaj) [reply src/middlewared/middlewared/restful.py:515]:
```
 if not method_args:
                # RFC 7231 specifies that a GET request can accept a payload body
                # This means that all the http methods now ( delete, get, post, put ) accept a payload body
```
Yes

Comment (themylogin) [reply src/middlewared/middlewared/restful.py:515]: That does not make it a good request semantic. You can accept any JSON variable just fine using query parameters `/items/query-options.extra.include_chart_schema=true&query-options.extra.retrieve_resources=false`

**Review by themylogin — CHANGES_REQUESTED** [review]

File: src/middlewared/middlewared/restful.py [hunk 2]
```diff
@@ -227,7 +227,12 @@ def add_path(self, path, operation, methodname, service_config):
                         'schema': {'type': 'string'},
                     },
                 ]
-            elif accepts and not (operation == 'delete' and method['item_method'] and len(accepts) == 1):
+            elif '{id}' in path and method['filterable']:
+                desc = f'{desc}\n\n' if desc else ''
+                opobject['description'] = desc + '`query-options.extra` can be specified as query parameters.'
```
Comment (themylogin) [inline src/middlewared/middlewared/restful.py:232]: Can't they be specified for e.g. `GET /storage/pool/`, why a special case?

**Review by themylogin — APPROVED** [review]
Thank you!

File: src/middlewared/middlewared/restful.py [hunk 3]
```diff
@@ -497,9 +518,13 @@ def get_post_args(self, req, resp, **kwargs):
                     filterid = kwargs['id']
                     if filterid.isdigit():
                         filterid = int(filterid)
+                    extra = {}
+                    for key, val in list(req.query.items()):
```
Comment (themylogin) [inline src/middlewared/middlewared/restful.py:522]: Should it be `extra.` here too?

**Review by themylogin — CHANGES_REQUESTED** [review]

File: src/middlewared/middlewared/restful.py (hunk 3 above)
Comment (sonicaj) [reply src/middlewared/middlewared/restful.py:522]: I don't think so because we are not expecting anything else here ?

**Review by themylogin — APPROVED** [review]

**Pull request merged**